[Test] Fix CRUDDocumentationIT

tlrx opened the pull request:
Similarly to other documentation tests in the high level client, the
asynchronous operation like update, index or delete can make the test
fail if it sneak in between the middle of another operation.

This commit moves the async doc tests to be the last ones executed and
adds await busy loops to ensure that the asynchronous operations are
correctly terminated.

closes #28446

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/CRUDDocumentationIT.java:
```diff
@@ -240,9 +225,34 @@ public void onFailure(Exception e) {
+                } catch (IOException e) {
+                    return false;
+                }
+            }));
```
what happens if we don't wait for this one?

javanna replied:
I was also thinking, given that this is a recurring problem and we don't want to have latches to show in our docs, shall we make this more clear doing something through a listener that blocks (without showing in the snippets what the listener really does)?

tlrx replied:
> what happens if we don't wait for this one?

After each test method, the `ESRestTestCase.cleanUpCluster()` is executed. As you know this method deletes the cluster indices, waits for the pending cluster states tasks to finish and logs a message if there is still inflight requests. 

By not waiting for the async request to respond, it's possible that an index got re created because the index (or update) operation is processed after the indices have been deleted and s recreates the index and breaks following tests. 

I'm pretty sure I saw this error, but sadly I can't find a test execution log that clearly shows this behavior. In any way I think it's better to wait for the completion of any write operation that could recreate an index for the reason I mentioned.

javanna replied:
makes sense, thanks for explaining.

javanna replied:
what do you think about the listener idea above?

tlrx replied:
I'd prefer to have a blocking listener than assertBusy loops but I guess we'll have to declare a simple listener just for the doc purpose like:

```
//tag::listener
ActionListener<IndexResponse> listener = new ActionListener<IndexResponse>() {
                @Override
                public void onResponse(IndexResponse indexResponse) {
                    // <1>
                }

                @Override
                public void onFailure(Exception e) {
                    // <2>
                }
            };
//end::listener
```

Then, out of documented snippet replace the listener by a blocking one:

```
// Replace the listener by a blocking one
final CountDownLatch latch = new CountDownLatch(1);
listener = new ActionListener<IndexResponse> {
.. use latch in on response/onfailure methods here
}
```

And finally show the async execution snippet using the listener reference:
```
// tag::index-execute-async
client.indexAsync(request, listener);
//end::async
```

Is that what you have in mind?

javanna replied:
yea this sounds good to me.

tlrx replied:
This will have an impact on documentation (2 code snippets instead of one), should we align all documentation for async execution to be like this? I think so.

javanna replied:
yes I would adapt the docs, let me know if you prefer me to take this over given that it's becoming a bigger task.

tlrx replied:
I can do it. Do you prefer a follow up PR or be part of this one?

javanna replied:
I don't mind either way

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/CRUDDocumentationIT.java:
```diff
@@ -497,9 +493,34 @@ public void onFailure(Exception e) {
+                } catch (IOException e) {
+                    return false;
+                }
+            }));
```
once you have changed the id of the update, it won't sneak in anymore and cause version conflict, so I think it doesn't necessarily needs to be moved down? But maybe it's good like you did. Maybe we don't even need to wait for it at this point?

tlrx replied:
That was for readability, as some test use the same document to index or update multiple times. But I can move them back if you prefer.

> Maybe we don't even need to wait for it at this point?

We need to wait  for any writing operation, but we can relax this for read operations.

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/CRUDDocumentationIT.java:
```diff
@@ -601,6 +608,34 @@ public void onFailure(Exception e) {
+                } catch (IOException e) {
+                    return false;
+                }
+            }));
```
same comment as above, once we change the id, do we need to still wait for it?

tlrx commented:
Thanks for the reviews @nik9000 and @javanna.

> Also, I don't think you had to move these to the last ones, right?

This is not a requirement of course but some tests use the same document (like /posts/doc/1) and update or index it multiple times. Since I changed the async operations to use a different doc I moved them down so that it does not interleave with other request executions as I found this more readable.
